evals/funnels/sql/steps.py

```diff
@@ -0,0 +1,222 @@
+    """
+    name = "no_table_halucinations_test"
+
+    if len(expected_tables) == 0 and len(tables_in_query) == 0:
+        # No tables expected, no tables generated.
+        return FunnelStepResult(name=name, passed=True)
+
+    # We want to check if each of these tables exist in the schema
+    for table_path in tables_in_query:
+        database_name, schema_name, table_name = parse_table_path(table_path)
+
+        # Check if database exists
+        if database_name not in schema:
+            return FunnelStepResult(
+                name=name,
+                passed=False,
+                notes=f"Database '{database_name}' does not exist in schema",
+            )
+
+        # Check if schema exists within database
+        if schema_name not in schema[database_name]:
+            return FunnelStepResult(
+                name=name,
+                passed=False,
+                notes=f"Schema '{schema_name}' does not exist in database '{database_name}'",
+            )
+
+        # Check if table exists within schema
+        if table_name not in schema[database_name][schema_name]:
+            return FunnelStepResult(
+                name=name,
+                passed=False,
+                notes=f"Table '{table_name}' does not exist in schema '{schema_name}' of database '{database_name}'",
+            )
+
+    return FunnelStepResult(name=name, passed=True)
+
+
+def no_column_table_mismatch_test(
+    table_details: List[TableDetails],
+    schema: Dict[str, Any],
+) -> FunnelStepResult:
+    """
+    Verifies that all columns referenced in the query exist in their respective tables.
+    This also doubles as a halucination test, since if a column doesn't exist in the table,
+    it must be a halucination.
+
+    Args:
+        table_details: Generated by the AI; list of table details, each containing a table name and a list of columns
+        schema: The database schema containing all valid tables and their columns
+    """
+    name = "no_column_table_mismatch_test"
+
+    for table_detail in table_details:
+        database_name, schema_name, table_name = parse_table_path(table_detail.name)
+        columns = table_detail.columns
+
+        # Our reference point:
+        # the columns that are actually in the schema
+        columns_in_schema = [
+            column["name"] for column in schema[database_name][schema_name][table_name]
+        ]
+
+        # Check if all columns exist in the table
+        for column in columns:
+            if column.startswith("*"):
+                # Skip if the column is a wildcard
+                continue
+            elif column not in columns_in_schema:
+                return FunnelStepResult(
+                    name=name,
+                    passed=False,
+                    notes=f"Column '{column}' not found in table '{table_name}'",
+                )
+
+    return FunnelStepResult(name=name, passed=True)
+
+
+def syntax_check_test(
```

Comment:
Let's call this `execute_without_errors_test` since it could fail for other reasons than the syntax not being correct